cluster/cluster.go

```diff
@@ -22,108 +26,104 @@ func validMode(m string) bool {
-}
-
-func AddPeer(remoteAddr *url.URL) {
-	mu.Lock()
-	peer := &Node{
-		remoteAddr:  remoteAddr,
-		stateChange: time.Now(),
+	log.Info("Starting cluster on %s:%d", cfg.BindAddr, cfg.BindPort)
+	list, err := memberlist.Create(cfg)
+	if err != nil {
+		log.Fatal(4, "Failed to create memberlist: "+err.Error())
 	}
-	peers = append(peers, peer)
-	go peer.Probe()
-	mu.Unlock()
-}
+	data, err := json.Marshal(Manager.ThisNode())
+	if err != nil {
+		log.Fatal(4, "Failed to marshal ThisNode metadata to json. %s", err.Error())
+	}
+	list.LocalNode().Meta = data
 
-func GetPeers() []*Node {
-	mu.Lock()
-	p := make([]*Node, len(peers))
-	copy(p, peers)
-	mu.Unlock()
-	return p
-}
+	Manager.SetList(list)
 
-func probePeers(interval time.Duration) {
-	ticker := time.NewTicker(interval)
-	for {
-		select {
-		case <-shutdown:
-			ticker.Stop()
-			return
-		case <-ticker.C:
-			mu.Lock()
-			for _, peer := range peers {
-				go peer.Probe()
-			}
-			mu.Unlock()
-		}
+	if peersStr == "" {
+		return
 	}
+	n, err := list.Join(strings.Split(peersStr, ","))
+	if err != nil {
+		log.Fatal(4, "Failed to join cluster: "+err.Error())
+	}
+	log.Info("joined to %d nodes in cluster\n", n)
 }
 
 // return the list of nodes to broadcast requests to
 // Only 1 peer per partition is returned. This list includes
 // ThisNode if it is capable of handling queries.
-func PeersForQuery() []*Node {
+func PeersForQuery() []Node {
+	thisNode := Manager.ThisNode()
 	// If we are running in single mode, just return thisNode
 	if Mode == ModeSingle {
-		return []*Node{ThisNode}
+		return []Node{thisNode}
 	}
 
-	peersMap := make(map[int32][]*Node)
-	if ThisNode.IsReady() {
-		for _, part := range ThisNode.GetPartitions() {
-			peersMap[part] = []*Node{ThisNode}
+	peersMap := make(map[int32][]Node)
+	if thisNode.IsReady() {
+		for _, part := range thisNode.Partitions {
+			peersMap[part] = []Node{thisNode}
 		}
 	}
-	mu.Lock()
-	for _, peer := range peers {
-		if !peer.IsReady() {
+
+	for _, peer := range Manager.PeersList() {
+		if !peer.IsReady() || peer.Name == thisNode.Name {
 			continue
 		}
-		for _, part := range peer.GetPartitions() {
+		for _, part := range peer.Partitions {
 			peersMap[part] = append(peersMap[part], peer)
 		}
 	}
-	mu.Unlock()
-	selectedPeers := make(map[*Node]struct{})
-	answer := make([]*Node, 0)
+
+	selectedPeers := make(map[string]struct{})
+	answer := make([]Node, 0)
 	// we want to get the minimum number of nodes
 	// needed to cover all partitions
 	for _, nodes := range peersMap {
 		selected := nodes[0]
 		// always prefer the local node which will be nodes[0]
 		// if it has this partition
-		if selected != ThisNode {
+		if selected.Name != thisNode.Name {
 			// check if we are already going to use one of the
 			// available nodes and re-use it
 			reusePeer := false
 			for _, n := range nodes {
-				if _, ok := selectedPeers[n]; ok {
+				if _, ok := selectedPeers[n.Name]; ok {
 					selected = n
 					reusePeer = true
```

Comment:
Couldn't this bit be shortened a little? 
If you give the outer `for` a label and just `continue` the outer `for` here, then you don't need the `reusePeer` flag and the two `if`s on line `134` and `139`